internal/app/referenceclient/wire_tracer.go

```diff
@@ -0,0 +1,159 @@
+// Copyright 2023-2024 The Connect Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package referenceclient
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"strings"
+	"sync"
+
+	"connectrpc.com/conformance/internal"
+	v1 "connectrpc.com/conformance/internal/gen/proto/go/connectrpc/conformance/v1"
+	"connectrpc.com/conformance/internal/tracer"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+type wireKey struct{}
+
+// WireDetails encapsulates the wire details to track for a roundtrip.
+type WireDetails struct {
+	// The actual HTTP status code observed.
+	StatusCode int32
+	// The actual trailers observed.
+	Trailers []*v1.Header
+	// The actual JSON observed on the wire in case of an error from a Connect server.
+	// This will only be non-nil if the protocol is Connect and an error occurred.
+	ConnectErrorRaw *structpb.Struct
```

Comment:
Aside: I got another one we'll want to capture. Not that you need to add it here; I can come back and add it soon.

But we will also want to capture the raw trailers in a gRPC-Web end-stream message. Because we'll want to validate that they were all lower-case on the wire.